### ISSUE TITLE
plain makefile not fully compatible with make_six

### ISSUE THREAD

**Issue by rdemaria** [SixTrack/SixTrack#37]

**Comment by vikasnt** [comment]
System : Ubuntu 14.04 on Oracle Virtual box 
Test used: bb bbe51 bbe52 bbe571ib0 bnl crabamp dipedge distance eric exact frs frs60 javier lost lostevery lostnotilt notilt  s316 thick4 tilt 

1.  First using make_six, generate the executable, and run the tests to ensure make_six is building fine. Output files from check_10 and check_90
[make_six_check90.txt](https://github.com/SixTrack/SixTrack/files/263216/make_six_check90.txt)
[make_six_check10.txt](https://github.com/SixTrack/SixTrack/files/263217/make_six_check10.txt)

2. Now using make file instead , again generated the executable, and run the test to compare with make_six for errors if any. Output files from check_10 and check_90
[make_check90.txt](https://github.com/SixTrack/SixTrack/files/263218/make_check90.txt)
[make_check10.txt](https://github.com/SixTrack/SixTrack/files/263219/make_check10.txt)
 Here,
check90 doesn't show any difference in result which means fort90 is correctly generated and tracking of particles is going correct even in make build.
check10 shows difference this time in test "eric,lostevery,s316 and thick4". Difference in fort10 but not in fort90 for these test probably means bug in code generated for post-processing of data done at the end of run.
Also, test "exact and bnl" failed due to missing fort.10 file but this doesn't really reflect on make file until we test after providing missing fort.10.

**Comment by kyrsjo** [comment]
One issue we realized when discussing the make_six:

make_six creates a file called "Makefile" (uppercase M), which it later uses with 'make -f'. There is also the file "makefile" (lowercase letters), which is the new build system.

On operating systems with case sensitive file systems (Linux, optionally OS X) this works fine. On other operating systems, it would not work, one of the files will probably be overwritten.

Proposal: Rename the "Makefile" to "makefile_make_six", and rename the templates to "makefile_make_six.template" etc.. This should avoid any name collissions. @rdemaria , what do you think?

Please use "git mv" when renaming files.

**Comment by kyrsjo** [comment]
@rdemaria Also, I remember you had some ideas to make the new makefile (lowercase) Windows-compatible - could you remind us what this was?

**Comment by vikasnt** [comment]
--:Update on test done with current makefile:--

Test- prob1 and prob3,
Combined run time on Ubuntu 14.04 on Oracle Virtual box for make_six and make executable running in parallel on 2 virtual cpu: using 3GB RAM~24 hours

As expected make_six runs correctly and gives right output, while make runs wrong for both and fort.90 as well as fort.10 are different than canonical versions.

Another issue is RAM used by them while execution which for make is almost 600 MB while only 50 MB for make_six although they don't differ in terms of run time.
![ram usage](https://cloud.githubusercontent.com/assets/14842531/15299458/1c275fba-1bc2-11e6-8e11-0d925d164dd6.png)

[make_check10.txt](https://github.com/SixTrack/SixTrack/files/266661/make_check10.txt)
[make_check90.txt](https://github.com/SixTrack/SixTrack/files/266663/make_check90.txt)
[make_six_check10.txt](https://github.com/SixTrack/SixTrack/files/266660/make_six_check10.txt)
[make_six_check90.txt](https://github.com/SixTrack/SixTrack/files/266662/make_six_check90.txt)

This also make earlier assumption that tracking is working perfectly to be wrong since we got fort.90 different this time.

**Comment by rdemaria** [comment]
The make file could be made simpler if we assume windows has a unix like
environment. The major complication is passing the flags into the ast
files. I think this could be one problem. The other problem could be
different compilation flags for crlibm.

--:Update on test done with current makefile:--

Test- prob1 and prob3,
Combined run time on Ubuntu 14.04 on Oracle Virtual box for make_six and
make executable running in parallel on 2 virtual cpu: using 3GB RAM~24 hours

As expected make_six runs correctly and gives right output, while make runs
wrong for both and fort.90 as well as fort.10 are different than canonical
versions.

Another issue is RAM used by them while execution which for make is almost
600 MB while only 50 MB for make_six although they don't differ in terms of
run time.
[image: ram usage]
<https://cloud.githubusercontent.com/assets/14842531/15299458/1c275fba-1bc2-11e6-8e11-0d925d164dd6.png>

make_check10.txt
<https://github.com/SixTrack/SixTrack/files/266661/make_check10.txt>
make_check90.txt
<https://github.com/SixTrack/SixTrack/files/266663/make_check90.txt>
make_six_check10.txt
<https://github.com/SixTrack/SixTrack/files/266660/make_six_check10.txt>
make_six_check90.txt
<https://github.com/SixTrack/SixTrack/files/266662/make_six_check90.txt>

This also make earlier assumption that tracking is working perfectly to be
wrong since we got fort.90 different this time.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/SixTrack/SixTrack/issues/37#issuecomment-219506427>

**Comment by kyrsjo** [comment]
Crazy idea: Could we do the windows compile with MinGW?

**Comment by rdemaria** [comment]
Yes, this is what eric is trying to do. Do you have access to Laurent's
windows machine. He has everything set up. The unix makefile almost works...

**Comment by vikasnt** [comment]
After going through the current makefile, we believe following issues need to be addressed:

1. Valid combination checking code seems to be not used : 
options dependencies (+)enable, -disable, !exclude
beamgas    := +collimat
boinc      := +cpss
bpm        := !da !collimat
collimat   := -crlibm !da
cpss       := +cr !da !collimat
da         := -naglib !collimat
hdf5       := +collimat
windows    := +boinc +crlibm 
Maybe it is still being done in some default way, but might give error when user chooses some option(s)?

2. vvector option is not used in makefile which is by default ON is make_six, so after switching it ON running test to understand how it affects current difference in sixtrack executable.

3. hugenblz option should be turned ON if user asks, and by default should be OFF. This might explain large RAM used in test of makefile based executable.

**Comment by kyrsjo** [comment]
Thanks Vikas! Please let us know when the tests are finished.

@rdemaria , @ericmcintosh : Do you know what vvector is? Is it correct that it is on, or should it be off? Any code which is there, but is actually obsolete (such as the "+if .not. vvector" code)?

@rdemaria @ldeniau : No, I don't have access to his machine. Let's see where we get after Vikas has fixed the bugs in the Unix makefile; What he mentions as point 1 seems to be a pretty nasty bug, the option checking is simply never done. It looks like this code was never finished... It may be useful to to get the printout of what is going wrong with the build there? Oh, and why is the minGW build of @ericmcintosh failing?

**Comment by vikasnt** [comment]
Update:
makfile seems almost correct for linux(32 and 64 bit) and mac(64bit) and now building on mac 64 bit can be done and the executable passes all test.
Still one issue which actually belongs to tests and not makefile itself remains which i believe @ericmcintosh can help out with?
On running check_10 and check_90 in mac, it gives error related to format of binary files. Below is an example for "prob1" test with run no. 641

./check_10 641
++++++++++++++++++++++++++++++++++++++
Comparing canonical to run_pro_sixtrack_641_2806
++++++++++++++++++++++++++++++++++++++
      ====================================
      CASE prob1
      ====================================
/Users/maezono/Desktop/Gsoc/github/SixTrack/SixTest/bin/mychecktens: line 5: /Users/maezono/Desktop/Gsoc/github/SixTrack/SixTest/bin/checkf10: cannot execute binary file

./check_90 641
++++++++++++++++++++++++++++++++++++++
Comparing canonical to run_pro_sixtrack_641_2806
++++++++++++++++++++++++++++++++++++++
prob1
      ====================================
      CASE prob1
      ====================================
./check_90: line 63: /Users/maezono/Desktop/Gsoc/github/SixTrack/SixTest/bin/read90: cannot execute binary file
./check_90: line 65: /Users/maezono/Desktop/Gsoc/github/SixTrack/SixTest/bin/read90: cannot execute binary file
fort.90 SAME

Although on running check on test results produced by mac executable in linux(Ubuntu) , both check_10 and check_90 run successful.

**Comment by ericmcintosh** [comment]
Hi Vikas, I'll get back to you tomorrow latest.
In the meantime I suggest that the binaries of
of these programs are for Linux. They need to be rebuilt
for MacOS. read90 has it's own directory in try and others
can be found in bin in try. All sounds good, and I usually
check on Linux anyway. Still it should be possible on
MacOS. If not documented, I shall do so. I am trying to
clean up a big mess with this anyway.  Thanks.  Eric.

**Comment by kyrsjo** [comment]
I think we can close this now? Regarding read90, I ended up doing a few things to that code - Eric I can show you next time we see each other. The windows issues are tracked in issue #47 .
